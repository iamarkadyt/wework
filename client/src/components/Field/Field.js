import React from 'react'
import DropdownList from 'react-widgets/lib/DropdownList'
import DateTimePicker from 'react-widgets/lib/DateTimePicker'
import Multiselect from 'react-widgets/lib/Multiselect'
import Moment from 'moment'
import momentLocalizer from 'react-widgets-moment'
import './Field.scss'
import { node, string, func, object, bool, number, arrayOf, oneOfType, instanceOf } from 'prop-types'

Moment.locale('en')
momentLocalizer()

const field = ({
    // different fields require different props
    type,
    name, value, onChange, // all except submit and button
    placeholder, error, // all except checkbox, submit and button
    label, // checkbox, submit, button
    list, // list, multiselect
    rows, // textarea
    disabled, // date
    onClick, style, // button, linkButton, textarea
    inline, // container
    containerStyle, // container
    children
}) => {
    let field = null
    const commonProps = { name, id: name, value, onChange }

    switch (type) {
        case "textarea":
            field = <textarea
                {...commonProps}
                placeholder={placeholder || 'Say something...'}
                className="rw-widget-container rw-widget-input"
                style={{
                    padding: '.5rem',
                    fontSize: '1rem',
                    width: '100%',
                    resize: 'none',
                    ...style
                }}
                rows={rows} />
            break
        case "date":
            field = <DateTimePicker
                {...commonProps}
                placeholder={placeholder || 'Pick a date...'}
                time={false}
                format="MMM DD, YYYY"
                editFormat="M/DD/YYYY"
                disabled={disabled} />
            break
        case "list":
            field = <DropdownList
                {...commonProps}
                placeholder={placeholder || "Choose one..."}
                data={list} />
            break
        case "multiselect":
            field = <Multiselect
                {...commonProps}
                placeholder={placeholder || "Enter a few items..."}
                data={list} />
            break
        case "checkbox":
            field = <label htmlFor={name}>
                <input type="checkbox" {...commonProps} />
                &nbsp;&nbsp;{label}
            </label>
            break
        case "submit":
            field = <button
                type="submit"
                className="button-action">
                {label || children}
            </button>
            break
        case "button":
            field = <button
                className="button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        case "linkButton":
            field = <button
                className="link-button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        default:
            field = <input
                {...commonProps}
                placeholder={placeholder}
                className="rw-widget-container rw-widget-input"
                type={type} />
    }

    return (
        <div style={containerStyle}
            className={['Field-container', inline ? 'inline' : ''].join(' ')}>
            {label
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <label htmlFor={name}>{label}</label>}
            {field}
            {error
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <span className="error">{error}</span>}
        </div>
    )
}

field.propTypes = {
  type: string,
  name: string,
  value: oneOfType([string, arrayOf(string), bool, instanceOf(Date)]),
  onChange: func,
  placeholder: string,
  error: string,
  label: string,
  list: arrayOf(string),
  rows: oneOfType([string, number]),
  disabled: bool,
  onClick: func,
  style: object,
  inline: bool,
  containerStyle: object,
  children: oneOfType([arrayOf(node), node])
}

export default field
