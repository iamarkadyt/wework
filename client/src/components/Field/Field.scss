@import 'variables.scss';
.Field-container {
    &.inline {
        display: inline;
    }
    width: 100%;
    margin: 1rem 0 0 0;
    input[type=checkbox] {
        transform: scale(1.5);
    }
    .button,
    button[type=submit] {
        min-width: 6rem;
        border-radius: .2rem;
        font-size: 1.1rem;
        padding: .2rem 1rem;
        border-color: $color-button-border;
        background: $color-button-backround;
        color: $color-button-font;
        &:active {
            border-color: $color-button-active-border;
            background: $color-button-active-background;
            color: $color-button-active-font;
        }
        &:hover:not(:active) {
            border-color: $color-button-hover-border;
            background: $color-button-hover-backround;
            color: $color-button-hover-font;
        }
    }
    button[type=submit] {
        float: right;
    }
    .button-action {
        background-color: darken(cornflowerblue, 10%);
        border-color: darken(cornflowerblue, 10%);
        background: linear-gradient(to bottom, darken(cornflowerblue, 10%), darken(cornflowerblue, 20%));
        color: white;
        &:active {
            background: linear-gradient(to bottom, darken(cornflowerblue, 25%), darken(cornflowerblue, 15%));
        }
        &:hover:not(:active) {
            background: linear-gradient(to bottom, darken(cornflowerblue, 8%), darken(cornflowerblue, 18%));
        }
    }
    .link-button {
        background: none;
        border: none;
        font-weight: bold;
        font-size: 1.1rem;
        color: $color-font-secondary;
        &:hover {
            color: $color-font-main;
            cursor: pointer;
        }
    }
    input[type=text],
    input[type=password] {
        font-size: 1rem;
        padding: .6rem;
        width: 100%;
    }
    label {
        display: inline-block;
        padding: .2rem;
    }
    textarea {
        background-color: $color-field-background;
        border-color: $color-field-border;
        color: $color-field-font;
    }
    .error {
        display: inline-block;
        color: $color-font-error;
        padding: .2rem;
    }
}