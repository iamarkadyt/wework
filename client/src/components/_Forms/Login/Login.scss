@import 'mixins.scss';
.Login-container {
    width: 100%;
    height: 100%;
    @include flexList();
    form {
        @include flexList();
        width: 30rem;
    }
}