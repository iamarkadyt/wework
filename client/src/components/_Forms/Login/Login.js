import React from 'react'
import { connect } from 'react-redux'
import Field from '../../Field/Field'
import { Link, withRouter } from 'react-router-dom'
import './Login.scss'
import {
    loginUser,
    fetchFollowers,
    fetchSubscriptions,
} from '../../../state/actions/userActions'
import { fetchUsersProfile } from '../../../state/actions/profileActions'
import { func, object } from 'prop-types'

class Login extends React.Component {
    state = {
      email: '',
      password: ''
    }

    render() {
        return <div className="Login-container">
            <form onSubmit={e => {
                e.preventDefault()
                this.props.loginUser(this.state, userId => {
                    this.props.fetchSubscriptions(userId)
                    this.props.fetchFollowers(userId)
                    this.props.fetchUsersProfile()
                    this.props.history.push('/feed')
                })
            }}>
                <Field
                    type="text"
                    name="email"
                    value={this.state.email}
                    onChange={e => this.setState({ email: e.target.value })}
                    label="Email:"
                    error={this.props.errors.email} />
                <Field
                    type="password"
                    name="password"
                    value={this.state.password}
                    onChange={e => this.setState({ password: e.target.value })}
                    label="Password:"
                    error={this.props.errors.password} />
                <div className="Login-buttons">
                    <span className="Login-options-info">
                        Don't have an account? <Link to="/signup">Sign Up</Link>
                    </span>
                    <Field
                        type="submit"
                        inline
                        containerStyle={{ margin: 0 }}
                        label="Log In" />
                </div>
            </form>
        </div>
    }
}

Login.propTypes = {
  loginUser: func.isRequired,
  fetchFollowers: func.isRequired,
  fetchSubscriptions: func.isRequired,
  fetchUsersProfile: func.isRequired,
  errors: object.isRequired
}

export const _UnconnectedLogin = Login
export default withRouter(connect(state => ({
    errors: state.err.formErrors
}), { loginUser, fetchFollowers, fetchSubscriptions, fetchUsersProfile })(Login))
