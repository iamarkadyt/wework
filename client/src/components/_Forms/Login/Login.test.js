/**
 * Contract:
 *
 * What my component renders?
 * (snapshot testing)
 * - form containing 2 fields
 *   1 div with 1 submit field and 1 text span
 *
 * What do I pass to components I render?
 * (props passing)
 * Look at:
 * - form
 * - email field
 * - password field
 * - submit field
 * - Link component
 *
 * What effects do the props I receive have?
 * (props effects)
 * - subfields in [errors] object cause Field components to show error spans right below input fields
 * - vice versa is true: no errors, no red texts
 *
 * How do I use functions (any of them)?
 * (functions)
 * - 4 thunk functions must be called with correct args
 * - onSubmit must preventDefault(), call all 4 thunk actions and history.push()
 * - Fields' onChange functions receive event obj and call setState correctly
 *
 * What does the state hold?
 * (state)
 * - email textfield data
 * - password field data
 *
 * How and When do I update/invalidate state? 
 * (state)
 * - When user types, respective fields update state through setState calls
 * - No invalidation occurs throughout the whole lifecycle of the component
 *
 * How state affects other components?
 * (state)
 * - Updates text in text Fields
 *
 * Any context?
 * (context)
 * - react-router
 *   > history.push()
 *
 * Lifecycle hooks side effects?
 * - N/A
 *
 * Public ref API?
 * - N/A
 *
 * What happens when user interacts with my component?
 * (interaction)
 * - When user types, respective fields update state through setState calls
 * - When user clicks on [Submit] button form's onSubmit function is called
 * - When user clicks on signup <Link> it navigates to /signup
 */


import React from 'react'
import Login from './Login'
import Field from '../../Field/Field'
import { shallow } from 'enzyme'
import cloneDeep from 'lodash.clonedeep'

describe('Login', () => {
  let props
  let mountedLogin

  const getMockProps = () => {
    const actions = {
      fetchFollowers: jest.fn(),
      fetchSubscriptions: jest.fn(),
      fetchUsersProfile: jest.fn(),
      loginUser: jest.fn()
    }

    const history = {
      push: jest.fn()
    }

    return cloneDeep({
      ...actions,
      history,
      errors: {
        email: 'Email is wrong'
      }
    })
  }

  const loginComp = () => {
    if (!mountedLogin) {
      mountedLogin = shallow(<Login {...props} />)
    }
    return mountedLogin
  }

  beforeEach(() => {
    props = getMockProps()
    mountedLogin = undefined
  })
})
