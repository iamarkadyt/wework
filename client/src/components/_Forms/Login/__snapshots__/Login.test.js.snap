// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches snapshot 1`] = `
<div
  className="Login-container"
>
  <form
    onSubmit={[Function]}
  >
    <field
      label="Email:"
      name="email"
      onChange={[Function]}
      placeholder="Your email"
      type="text"
      value=""
    />
    <field
      label="Password:"
      name="password"
      onChange={[Function]}
      placeholder="Your password"
      type="password"
      value=""
    />
    <div
      className="Login-buttons"
    >
      <span
        className="Login-options-info"
      >
        Don't have an account? 
        <Link
          replace={false}
          to="/signup"
        >
          Sign Up
        </Link>
      </span>
      <field
        containerStyle={
          Object {
            "margin": 0,
          }
        }
        inline={true}
        label="Log In"
        name="submit"
        type="submit"
      />
    </div>
  </form>
</div>
`;
